pep-0646.rst

```diff
@@ -733,28 +733,166 @@ hopefully enable a thriving ecosystem of tools for analysing and verifying
 shape properties of numerical computing programs.
 
 
-Backwards Compatibility
-=======================
+Grammar Changes
+===============
+
+This PEP would require two grammar changes. Full diffs of ``python.gram``
+and simple tests to confirm correct behaviour are available at
+https://github.com/mrahtz/cpython/commits/pep646-grammar.
+
+Star expressions in indexes
+---------------------------
+
+The first grammar change enables use of star expressions in index operations (that is,
+within square brackets), necessary to support star-unpacking of TypeVarTuples.
+
+Before:
+
+::
+    
+    slices[expr_ty]:
+        | a=slice !',' { a }
+        | a[asdl_expr_seq*]=','.slice+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+
+After:
+
+::
+    
+    slices[expr_ty]:
+        | a=slice !',' { a }
+        | a[asdl_expr_seq*]=','.(slice | starred_expression)+ [','] { _PyAST_Tuple(a, Load, EXTRA) }
+
+As with star-unpacking in other contexts, the star operator would call ``__iter__`` on the callee,
+and add the contents of the resulting iterator to the list of index arguments.
+This would enable ``TypeVarTuple`` to be implemented as roughly:
+
+::
 
-In order to use the star operator for unpacking of ``TypeVarTuple`` instances,
-we would need to make two grammar changes:
+    class TypeVarTuple:
+        def __init__(self, name):
+            self._unpacked = _UnpackedTypeVarTuple(name) 
+        def __iter__(self):
+            yield self._unpacked
```

Comment:
Ideally I think dynamic checkers _would_ be able to recognise `_UnpackedTypeVarTuple`. I've added what I imagine `_UnpackedTypeVarTuple` to look like (and removed the underscore, since in fact it's meant to be visible to things outside `typing.py`) - is this enough?